service/matching/matching_engine.go

```diff
@@ -545,6 +569,42 @@ pollLoop:
 			workerVersionCapabilities: request.WorkerVersionCapabilities,
 			forwardedFrom:             req.GetForwardedSource(),
 		}
+		if e.config.EnableDeployments(req.PollRequest.Namespace) && req.PollRequest.WorkerVersionCapabilities.UseVersioning {
+			buildID := req.PollRequest.WorkerVersionCapabilities.BuildId
+			deploymentName := req.PollRequest.WorkerVersionCapabilities.DeploymentName
+
+			// check for duplicate request
+			dedupDeploymentKey := dedupDeploymentsKey{
+				taskQueueName: req.PollRequest.TaskQueue.Name,
+				taskQueueType: enumspb.TASK_QUEUE_TYPE_WORKFLOW,
+				buildID:       buildID,
+			}
+			if _, found := e.dedupDeployments[dedupDeploymentKey]; !found {
+
+				// signal input
+				updateDeploymentSignalInput := &deployspb.UpdateDeploymentSignalInput{
+					Name: req.PollRequest.TaskQueue.Name,
+					TaskQueueInfo: &deployspb.DeploymentWorkflowArgs_TaskQueueFamilyInfo_TaskQueueInfo{
+						TaskQueueType:   enumspb.TASK_QUEUE_TYPE_WORKFLOW,
+						FirstPollerTime: nil,
+					},
+				}
+
+				// workflow input
+				startDeploymentWorkflowArgs := &deployspb.DeploymentWorkflowArgs{
+					NamespaceName:     req.PollRequest.Namespace,
+					NamespaceId:       req.NamespaceId,
+					TaskQueueFamilies: nil,
+				}
+				_, err := e.startAndSignalDeploymentWorkflow(ctx, startDeploymentWorkflowArgs, updateDeploymentSignalInput, deploymentName, buildID, req.PollRequest.Identity)
```

Comment:
Two points here:

1. I am okay to change the naming of the signal although I do think we should have "deployment" in the signal naming for better readability (it's nit, can be ignored I believe)
2. I had considered using an `Update` but went ahead with signals because I thought we wanted to lazily initiate our workflow executions. Moreover, using `SignalWithStart` also provides less latency since it's async as opposed to sync write of an Update (yes, there is guarantee of the registration happening here). Overall, I thought both options have their pros and cons each but I thought having lazy initializations of workflows was something that we desired.

Reply:
For item, I think it's better to call the Update name `RegisterWorker` where "Worker" highlights that we are registering the set of TQ worker in a deployment. As oppose to future Update `RegisterBacklog` or similar when we see the first task (pinned) task for a versioned TQ.

(RegisterBacklog is a future thing that I anticipate we'd need to be able to build APIs for backlog stats for deployments, etc., no need to think about it right now)

> we should have "deployment" in the signal naming for better readability
If this helps with code readability, sure. As far as the wf is concerned all the signal names are scoped within a WF definition.